executor/index_lookup_hash_join.go

```diff
@@ -0,0 +1,397 @@
+	select {
+	case <-ctx.Done():
+		return true
+	case dst <- task:
+	}
+	return false
+}
+
+func (e *IndexNestedLoopHashJoin) newOuterWorker(innerCh chan *indexHashJoinTask) *indexHashJoinOuterWorker {
+	ow := &indexHashJoinOuterWorker{
+		outerWorker: outerWorker{
+			outerCtx:         e.outerCtx,
+			ctx:              e.ctx,
+			executor:         e.children[0],
+			executorChk:      chunk.NewChunkWithCapacity(e.outerCtx.rowTypes, e.maxChunkSize),
+			batchSize:        32,
+			maxBatchSize:     e.ctx.GetSessionVars().IndexJoinBatchSize,
+			parentMemTracker: e.memTracker,
+			lookup:           &e.IndexLookUpJoin,
+		},
+		innerCh: innerCh,
+	}
+	return ow
+}
+
+func (e *IndexNestedLoopHashJoin) newInnerWorker(taskCh chan *indexHashJoinTask, workerID int) *indexHashJoinInnerWorker {
+	// Since multiple inner workers run concurrently, we should copy join's indexRanges for every worker to avoid data race.
+	copiedRanges := make([]*ranger.Range, 0, len(e.indexRanges))
+	for _, ran := range e.indexRanges {
+		copiedRanges = append(copiedRanges, ran.Clone())
+	}
+	return &indexHashJoinInnerWorker{
+		innerWorker: innerWorker{
+			innerCtx:      e.innerCtx,
+			outerCtx:      e.outerCtx,
+			ctx:           e.ctx,
+			executorChk:   chunk.NewChunkWithCapacity(e.innerCtx.rowTypes, e.maxChunkSize),
+			indexRanges:   copiedRanges,
+			keyOff2IdxOff: e.keyOff2IdxOff,
+		},
+		taskCh:            taskCh,
+		joiner:            e.joiner,
+		joinChkResourceCh: e.joinChkResourceCh[workerID],
+		resultCh:          e.resultCh,
+		matchedOuterPtrs:  make([][]byte, 0, 8),
+	}
+}
+
+func (iw *indexHashJoinInnerWorker) run(ctx context.Context) {
+	var task *indexHashJoinTask
+	ok, joinResult := iw.getNewJoinResult(ctx)
+	if !ok {
+		return
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case task, ok = <-iw.taskCh:
+		}
+		if !ok {
+			break
+		}
+		if task.err != nil {
+			joinResult.err = task.err
+			break
+		}
+		err := iw.handleTask(ctx, task, joinResult)
+		if err != nil {
+			joinResult.err = err
+			break
+		}
+	}
+	if joinResult.err != nil || (joinResult.chk != nil && joinResult.chk.NumRows() > 0) {
+		select {
+		case iw.resultCh <- joinResult:
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (iw *indexHashJoinInnerWorker) getNewJoinResult(ctx context.Context) (bool, *indexHashJoinResult) {
+	joinResult := &indexHashJoinResult{
+		src: iw.joinChkResourceCh,
+	}
+	ok := true
+	select {
+	case joinResult.chk, ok = <-iw.joinChkResourceCh:
+	case <-ctx.Done():
+		return false, nil
+	}
+	return ok, joinResult
+}
+
+func (iw *indexHashJoinInnerWorker) buildHashTableForOuterResult(task *indexHashJoinTask) (err error) {
+	keyBuf := make([]byte, 0, 64)
+	valBuf := make([]byte, 8)
+	for rowIdx, numRows := 0, task.outerResult.NumRows(); rowIdx < numRows; rowIdx++ {
+		var hasNull bool
+		if task.outerMatch != nil && !task.outerMatch[rowIdx] {
```

Comment:
No need to consider `outerMatch` when building the hash table?

Reply:
The unmatched outer rows can be called with `onMissMatch`,
we consider the `outerMatch` here to save the `map.Get` and `map.Put` overhead.